patch-db/src/locker/trie.rs

```diff
@@ -449,16 +449,52 @@ impl LockTrie {
             Some((first, rest)) => match self.children.get(first) {
                 None => (vec![&self.state], None),
                 Some(t) => {
-                    let (mut v, t) = t.ancestors_and_trie(&rest);
+                    let (mut v, t) = t.ancestors_and_trie_json_path(&rest);
                     v.push(&self.state);
                     (v, t)
                 }
             },
         }
     }
+    fn ancestors_and_trie_model_paths<'a>(
+        &'a self,
+        path: &[JsonGlobSegment],
+    ) -> (Vec<&'a LockState>, Option<&'a LockTrie>) {
+        let head = path.get(0);
+        match head {
+            None => (Vec::new(), Some(self)),
+            Some(JsonGlobSegment::Star) => {
+                let mut v = Vec::new();
+                let mut t = Some(self);
+                for lock_trie in self.children.values() {
+                    let (mut v_, t_) = lock_trie.ancestors_and_trie_model_paths(&path[1..]);
+                    v.append(&mut v_);
+                    t = t_.or(t);
+                }
+                (v, t)
+            }
+            Some(JsonGlobSegment::Path(x)) => match self.children.get(x) {
+                None => (vec![&self.state], None),
+                Some(t) => {
+                    let (mut v, t) = t.ancestors_and_trie_model_paths(&path[1..]);
+                    v.push(&self.state);
+                    (v, t)
+                }
+            },
+        }
+    }
```

Comment:
It's not enough to have this alone...or'ing the options together won't capture the situation where both tries are valid. This will break bookkeeping guaranteed.